keps/sig-network/20200406-graceful-termination-external-traffic-policy-local.md

```diff
@@ -0,0 +1,190 @@
+when the node is decommissioned from the loadbalancer's node pool.
+
+## Motivation
+
+### Goals
+
+* enable zero downtime rolling updates for Services with ExternalTrafficPolicy=Local via node ports or loadbalancer IPs.
+
+### Non-Goals
+
+* changing the behavior of terminating pods/endpoints outside the scope of Services with ExternalTrafficPolicy=Local via a nodeport/loadbalancerIPs/externalIPs.
+
+## Proposal
+
+### User Stories [optional]
+
+#### Story 1
+
+As a user I would like to do a rolling update of a Deployment fronted by a Service Type=LoadBalancer with ExternalTrafficPolicy=Local.
+If a node that has only 1 pod of said deployment goes into the `Terminating` state, all traffic to that node is dropped until either a new pod
+comes up or my cloud provider removes the node from the loadbalancer's node pool. Ideally the terminating pod should gracefully handle traffic to this node
+until one of the above conditions are satifised.
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Additions to EndpointSlice
+
+Today, terminating endpoints are treated the same as "not ready" endpoints. In order to make a distinction between an endpoint that is terminating
+and an endpoint that has failed readiness checks, a `Terminating` field would be added to `EndpointCondition` as part of the EndpointSlice API.
+
+No additions will be made to Endpoints meaning this change will depend on the EndpointSlice API being enabled.
+
+#### Endpoint Handling in kube-proxy
+
+To enable zero downtime deployments for Services with ExternalTrafficPolicy=Local, kube-proxy must also be aware of all
+terminating endpoints. To accomplish this, kube-proxy's service/endpoints caching would also collect any endpoint with `Terminating=true` in its
+`EndpointCondition`.
+
+The `Endpoint` interface as part of kube-proxy's endpoint cache would include a new method `IsTerminating()` which returns true if an endpoint is terminating
+according to the EndpointSlice API.
+```go
+// Endpoint in an interface which abstracts information about an endpoint.
+type Endpoint interface {
+	// String returns endpoint string.  An example format can be: `IP:Port`.
+	// We take the returned value as ServiceEndpoint.Endpoint.
+	String() string
+	// IsTerminating returns true if the endpoint is terminating (pod.DeletionTimestamp != nil)
+	IsTerminating() bool
+	// GetIsLocal returns true if the endpoint is running in same host as kube-proxy, otherwise returns false.
+	GetIsLocal() bool
+	// GetTopology returns the topology information of the endpoint.
+	GetTopology() map[string]string
+	// IP returns IP part of the endpoint.
+	IP() string
+	// Port returns the Port part of the endpoint.
+	Port() (int, error)
+	// Equal checks if two endpoints are equal.
+	Equal(Endpoint) bool
+}
+```
+
+In the event that all local endpoints (`GetIsLocal() == true`) are terminating (`IsTerminating() == true`), kube-proxy should direct all traffic for the Service's
```

Comment:
I think this is the part I am unsure of.

Consider the full space of { ready/not_ready, terminating/not_terminating }.

If we have any pod that is ready (R), we must not consider not_ready (!R) pods or terminating (T) pods at all.  If we have 0 ready pods, we may consider terminating pods.

I know for a fact that some pods start failing readiness upon SIGTERM (so they can start upstream LBs removal process ASAP).  I also know for a fact that most pods do not start failing readiness upon SIGTERM.

So if you consider {terminating, ready} different from {terminating, not_ready}, we don't actually solve this problem completely.  The logic has to have 3 levels:

1) if any {R, !T} pods, use those
2) else if any {R, T} pods, use those
3) else if any {!R, T} pods, use those

At that point you might as well add step 4 and fall back on {!R, !T} pods when that's all that you have.  It's probably not worse than a blackhole.

So, does "terminating" actually matter as a flag?  Or can we flatten (!R || T) => !R ?  The only time it would matter is if you have some {R, T} pods and some {!R, !T} pods AND NOTHING ELSE.  When could that happen?  Maybe if you have a small-N deploy that had a bad push.  Eg. 2 replicas, 1 has restarted and is crash-looping, the other has already been SIGTERMed.  The 2nd replica COULD server good responses, but it's going down either way..

From a performance POV - treating these as 4 states potentially almost doubles the number of writes to EndpointSlices.  Today we go {!R}->{R}->{} (3 writes).  After this we go  {!R, !T}->{R, !T}->{R, T}->{!R, !T}->{} (5 writes).  I know this is of concern to sig-scalability (@wojtek-t).

Simpler might be to ignore readiness when terminating, so {!R, !T}->{R, !T}->{x, T}->{} (4 writes).  It's less correct in that one corner case, but I am not sure it's worth the cost to fix.  Thoughts?

Reply:
>  So if you consider {terminating, ready} different from {terminating, not_ready}, we don't actually solve this problem completely. The logic has to have 3 levels:

> 1.  if any {R, !T} pods, use those
> 2.  else if any {R, T} pods, use those
> 3.  else if any {!R, T} pods, use those

> At that point you might as well add step 4 and fall back on {!R, !T} pods when that's all that you have. It's probably not worse than a blackhole

After some more thought I actually think it would be more "correct" as the following (really only point 3. is different): 

1) if {R, !T} > 0, send traffic there
2) if {R, !T} == 0 but {R, T} > 0, send traffic to {R, T}
3) if all endpoints are {!R, T} OR {!R, !T}, drop all traffic.

Sending traffic to any pod that fails readiness seems like we're breaking API contract. I'm personally in favor of only falling back to {R, T} endpoints as it would also encourage users to use readiness probes correctly. But... I agree it may not be very pragmatic because of some of the scalability / complexity concerns you mentioned. 

> Simpler might be to ignore readiness when terminating, so {!R, !T}->{R, !T}->{x, T}->{} (4 writes). It's less correct in that one corner case, but I am not sure it's worth the cost to fix. Thoughts?

I think this is likely what we'll have to do if scalability is a blocker? I'll wait for @wojtek-t to comment :).

Reply:
Ignoring my scalability hat for a moment, I personally think that I would *not* try to distinguish between {R, T} and {!R, T}. The reason is that we also have delay there (before kubelet observes !R and reports it, before controllers processes it and updates Endpoint/EndpointSlice, and before it's reflected in network (iptables or whatever)). So even though it's reported as {R, T}, it may already be in {!R, T} state.
The {!R, !T} -> {R, !T} is different, because we're risking being lucky, not unlucky.

So personally I wouldn't consider `{!R, !T}->{R, !T}->{R, T}->{!R, !T}->{} (5 writes).` even ignoring my scalability hat, and would just limit ourselves to ` {!R, !T}->{R, !T}->{x, T}->{}`.

Going further to what Andrew wrote - in my personal opinion this would make sense if we the states would really be as reported. But we only have eventual consistency, as I mentioned above.


Now putting my scalability hat back. Obviously the less writes the better. But we're also  not introducing new workflows here, we're just increasing a constant. Even going from 3 to 4 will have non-negligible impact and we will need test that. Though, there are still things we can do. If possible, I would try hard to avoid going from 3 to 5, but I don't think we can with confidence give answers that 3->5 won't work and 3->4 will work. Both may require some other kinds improvements (though at different levels) and we won't know until we really test that. But I hope with 3->4 it may just work.

@thockin @andrewsykim - thoughts?

Reply:
I think that makes sense for the majority of cases, but I think there are pods that intentionally fail readiness probes on SIGTERM as an important part of their business logic (not just to fail  LB checks). What would be the consequences of sending traffic to those endpoints even though their readiness probe fails? My guess for the worse case is somewhere up to 10s - 15s -- it largely depends on the load balancer's health check interval. 

I think communicating this behavior to users gets a little weird too. We would have to say something like:

> "The kubelet uses readiness probes to know when a container is ready to start accepting traffic, but there are certain scenarios with Services using externalTrafficPolicy=Local where this is not respected". 

If we only fallback to **ready** terminating endpoints, we're not breaking existing contracts/behaviors when falling back to those ready endpoints (assuming that always excluding terminating endpoint is a "bug" and not a feature today).

> Now putting my scalability hat back. Obviously the less writes the better. But we're also not introducing new workflows here, we're just increasing a constant. Even going from 3 to 4 will have non-negligible impact and we will need test that. Though, there are still things we can do. If possible, I would try hard to avoid going from 3 to 5, but I don't think we can with confidence give answers that 3->5 won't work and 3->4 will work. Both may require some other kinds improvements (though at different levels) and we won't know until we really test that. But I hope with 3->4 it may just work.

It's worth clarifying that 3 -> 5 writes assumes that every terminating endpoint goes from {R, T} -> {!R, T}. But I think in reality most pods will just go from {R, T} -> {} and skip the !R state in the middle. So even if we only fallback to {R, T} endpoints, it'll still be roughly 4 writes on average. I'm basing this off of personal experience and what Tim said earlier about most pods not failing readiness on SIGTERM today. Please correct me if the assumption is wrong.

Reply:
> there are pods that intentionally fail readiness probes on SIGTERM as an important part of their business logic

Do they stop accepting or just fail readiness?

So I know that there are pods that "artificially" fail readiness and you know there are pods that "meaningfully" fail readiness and we need to handle both.

So that leaves us with a worst-case Endpoints object that holds (over time):
   {} (no pods)  
->{!R, !T} (all pods are unready, probably starting up, P3)
->{R, !T} (at least one pod is ready and running normally, P0)
->{R, T} (all pods are terminating, at least one is still ready, P1) 
->{!R, T} (all pods are terminating and not ready, P2)
->{} (no pods)

{!R, !T} is not very interesting though realistically it won't matter much because once any pod is R, those will be ignored.  So we generally will only see a few of those before the first ones become ready.  For what it's worth, this is what Google's LB does - if it can't find any healthy endpoints, it uses unhealthy ones, on the assumption that it will never be worse than nothing, and sometimes it will be better. 

I also buy Andrew's point that {!R, T} is not super common, so probably won't factor a ton.

I would suggest we explicitly support all states and then see how that impacts scalability.  If it is dire we can consider dropping {!R, !T} and/or flattening {x, T}.  I do not think we can totally ignore {!R, T}.

Are we SURE that kubelet even keeps running readiness probes after it sees `deletionTimestamp` ?  This could all be moot - {!R, T} may not exist. :)

Reply:
> Are we SURE that kubelet even keeps running readiness probes after it sees deletionTimestamp ? This could all be moot - {!R, T} may not exist. :)

My assumption is yes because prior to https://github.com/kubernetes/kubernetes/pull/89056, epslice tracked ready state for terminating endpoints. I will do some extra testing to make sure  before making more updates here.

Reply:
Extra validation in https://github.com/kubernetes/kubernetes/pull/89780#issuecomment-618108063 that EPSlice can track {R,T} and {!R,T} endpoints, no changes in kubelet required.

Reply:
OK - I buy those points, thanks! [It would be good to actually add this discussion to the KEP - it might be useful for future readers (or future me :-) )

> I would suggest we explicitly support all states and then see how that impacts scalability. If it is dire we can consider dropping {!R, !T} and/or flattening {x, T}. I do not think we can totally ignore {!R, T}.

That sounds reasonable.